openedx/core/djangoapps/monitoring_utils/middleware.py

```diff
@@ -72,3 +73,83 @@ def process_exception(self, request, exception):  # pylint: disable=unused-argum
         """
         self._batch_report()
         return None
+
+
+class MonitoringMemoryMiddleware(object):
+    """
+    Middleware for monitoring memory usage.
+    """
+    memory_data_key = u'memory_data'
+
+    def process_request(self, request):
+        log_prefix = self._log_prefix(u"Before", request)
+        self._cache[self.memory_data_key] = self._memory_data(log_prefix)
+
+    def process_response(self, request, response):
+        log_prefix = self._log_prefix(u"After", request)
+        new_memory_data = self._memory_data(log_prefix)
+
+        log_prefix = self._log_prefix(u"Diff", request)
+        self._diff_memory_data(log_prefix, new_memory_data, self._cache.get(self.memory_data_key))
+        return response
+
+    @property
+    def _cache(self):
+        """
+        Namespaced request cache for tracking memory usage.
+        """
+        return request_cache.get_cache(name='monitoring_memory')
+
+    def _log_prefix(self, prefix, request):
+        """
+        Returns a formatted prefix for logging for the given request.
+        """
+        return u"{} request '{} {}'".format(prefix, request.method, request.path)
+
+    def _memory_data(self, log_prefix):
+        """
+        Returns a dict with information for current memory utilization.
+        Uses log_prefix in log statements.
+        """
+        machine_data = psutil.virtual_memory()
+
+        process = psutil.Process()
+        process_data = {
+            'memory_info': process.get_memory_info(),
```

Comment:
psutil is only used here and one other place.  (instructor task).  Barring any conflicts in dependent packages, it should be fairly easy to upgrade.